src/peft/tuners/glora/layer.py

```diff
@@ -0,0 +1,335 @@
+                dtype=dtype,
+            )
+            B = self.prepare_path(
+                path_config["B"],
+                self.glora_Bd[active_adapter],
+                self.glora_Bu[active_adapter],
+                device=device,
+                dtype=dtype,
+            )
+            C = self.prepare_path(
+                path_config["C"],
+                self.glora_Cd[active_adapter],
+                self.glora_Cu[active_adapter],
+                device=device,
+                dtype=dtype,
+            )
+            D = self.prepare_path(path_config["D"], self.glora_D[active_adapter], device=device, dtype=dtype)
+            E = self.prepare_path(path_config["E"], self.glora_E[active_adapter], device=device, dtype=dtype)
+            weight_eff = self.weight + self.weight * A + B
+            bias_eff = self.bias
+            if bias_eff is not None:
+                bias_eff = bias_eff + bias_eff * D + E + torch.matmul(self.weight, C).squeeze(-1)
+            else:
+                new_bias_val = E + torch.matmul(self.weight, C).squeeze(-1)
+                if not torch.all(new_bias_val == 0):
+                    bias_eff = new_bias_val
+            result = F.linear(x, weight_eff, bias=bias_eff)
+        return result
+
+    def prepare_path(self, config: str, Xd: nn.Parameter, Xu: Optional[nn.Parameter] = None, device=None, dtype=None):
+        device = device or Xd.device
+        dtype = dtype or Xd.dtype
+        if Xu is not None:
+            if "LoRA" in config:
+                rank = int(config.split("_")[1])
+                X = torch.matmul(Xd[:, :rank], Xu[:rank, :])
+            elif "vector" in config:
+                X = Xd[:, 0].unsqueeze(1)
+            elif "constant" in config:
+                X = Xd[0, 0]
+            elif "none" in config:
+                X = torch.zeros(Xd.shape[0], Xu.shape[1], device=device, dtype=dtype)
+            else:
+                raise ValueError(f"Unknown config choice: {config} for decomposable path")
+        else:
+            if "vector" in config:
+                X = Xd
+            elif "constant" in config:
+                X = Xd[0]
+            elif "none" in config:
+                X = torch.zeros(1, device=device, dtype=dtype)
+            else:
+                raise ValueError(f"Unknown config choice: {config} for non-decomposable path")
+        return X.to(device=device, dtype=dtype)
+
+
+# Refactored GLoraLinear for PEFT compatibility
+class GLoraLinear(GLoraLayer, nn.Linear):
+    def __init__(self, in_features, out_features, bias=True, **kwargs):
+        nn.Linear.__init__(self, in_features, out_features, bias=bias)
+        GLoraLayer.__init__(self, in_features=in_features, out_features=out_features)
+        self.weight.requires_grad = False
+        if self.bias is not None:
+            self.bias.requires_grad = False
+        self._disable_adapters = False
+        self.active_adapters = []
+        self.merged_adapters = []
```

Comment:
Let's swap the inheritance order for `GloraLinear`. Also, I would refrain from using `nn.Linear` and inherit from `nn.Module` instead since a common usage pattern is to check for `nn.Linear` to distinguish between PEFT and non-PEFT methods.